crates/codegen/parser_generator/src/code_generator.rs

```diff
@@ -0,0 +1,303 @@
+    ScannerDefinitionRef, TriviaParserDefinitionRef,
+};
+
+use super::{
+    parser_definition::ParserDefinitionExtensions,
+    precedence_parser_definition::PrecedenceParserDefinitionExtensions,
+    scanner_definition::ScannerDefinitionExtensions, trie::Trie,
+};
+
+#[derive(Default, Serialize)]
+pub struct CodeGenerator {
+    referenced_versions: BTreeSet<Version>,
+
+    rule_kinds: BTreeSet<&'static str>,
+    token_kinds: BTreeSet<&'static str>,
+    production_kinds: BTreeSet<&'static str>,
+
+    top_level_scanner_names: BTreeSet<&'static str>,
+    scanner_functions: Vec<(&'static str, String)>, // (name of scanner, code)
+    scanner_contexts: Vec<ScannerContext>,
+
+    parser_functions: Vec<(&'static str, String)>, // (name of parser, code)
+
+    #[serde(skip)]
+    scanner_contexts_map: BTreeMap<&'static str, ScannerContext>,
+    #[serde(skip)]
+    all_scanners: BTreeMap<&'static str, ScannerDefinitionRef>,
+    #[serde(skip)]
+    current_context_name: &'static str,
+}
+
+#[derive(Serialize)]
+struct ScannerContext {
+    name: &'static str,
+    #[serde(skip)]
+    scanner_definitions: BTreeSet<&'static str>,
+    alpha_literal_scanner: String,
+    non_alpha_literal_scanner: String,
+    compound_scanner_names: Vec<&'static str>,
+}
+
+impl CodeGenerator {
+    pub fn write_source(output_dir: &PathBuf, grammar: Rc<Grammar>) -> Result<()> {
+        let mut code = Self::default();
+        grammar.accept_visitor(&mut code);
+        let code = &code;
+
+        let runtime_dir =
+            CargoWorkspace::locate_source_crate("codegen_parser_runtime")?.join("src");
+        let mut codegen = Codegen::read_write(&runtime_dir)?;
+
+        {
+            #[derive(Serialize)]
+            pub struct Template<'a> {
+                pub code: &'a CodeGenerator,
+            }
+            codegen.render(
+                Template { code },
+                runtime_dir.join("templates/kinds.tera"),
+                output_dir.join("kinds.rs"),
+            )?;
+        }
+
+        {
+            #[derive(Serialize)]
+            pub struct Template<'a> {
+                pub code: &'a CodeGenerator,
+                pub language_name: String,
+                pub versions: BTreeSet<Version>,
+            }
+            codegen.render(
+                Template {
+                    code,
+                    language_name: grammar.name.clone(),
+                    versions: grammar.versions.clone(),
+                },
+                runtime_dir.join("templates/language.tera"),
+                output_dir.join("language.rs"),
+            )?;
+        }
+
+        {
+            #[derive(Serialize)]
+            pub struct Template {}
+            codegen.render(
+                Template {},
+                runtime_dir.join("templates/mod.tera"),
+                output_dir.join("mod.rs"),
+            )?;
+        }
+
+        for file in &[
+            "cst.rs",
```

Comment:
instead of maintaining two lists and possibly falling out of sync, I suggest moving `crates/codegen/parser_runtime/src/*` into a sub-module, and then we can just copy the entire folder over?
Generated code can refer to the top-level new `mod.rs` instead of `lib.rs`.

Reply:
No, because some must not be copied, because they are there just to make the 'template' crate valid. e.g. `kinds.rs`.